File: emp_backend/src/main/java/com/example/demo/model/Report.java
package com.example.demo.model;

import jakarta.persistence.*;
import lombok.*;

@Getter
@Setter
@ToString
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Entity(name = "reports")
public class Report {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @ManyToOne
    @JoinColumn(name = "project_id")
    private Project project;

    @OneToOne
    @JoinColumn(name="employee_id")
    private Employee employee;

    private long totalHoursWorked;
    private double averagePerformance;
    private Task.PerformanceMeasure performanceMeasure;
    private long completionTimeVariance;
    private double efficiency;
    private int numberOfLeaves;
    private double averageDaysPerLeave;
    private Integer taskLengthInHours;
}

File: emp_frontend/src/app/models/report.ts
import { Employee } from "./employee";
import {Project} from "./project";

export class Report {
  id: number;
  employee?: Employee; // Optional reference to Employee model
  project?: Project;
  generatedDate: string; // ISO 8601 format (e.g., "2024-07-18")
  totalHoursWorked: number;
  averagePerformance: number;
  performanceMeasure!: 'EXCELLENT' | 'GOOD' | 'MEDIUM' | 'POOR'; // Enum values
  completionTimeVariance: number;
  efficiency: number;
  numberOfLeaves: number;
  averageDaysPerLeave: number;
  taskLengthInHours: number | null;

  constructor() {
    this.id = 0;
    this.project = undefined;
    this.employee = undefined;
    this.generatedDate = "";
    this.totalHoursWorked = 0;
    this.averagePerformance = 0;
    this.completionTimeVariance = 0;
    this.efficiency = 0;
    this.numberOfLeaves = 0;
    this.averageDaysPerLeave = 0;
    this.taskLengthInHours = null;
  }
}
